SmartPointers/shared_pointer.hpp

```diff
@@ -1,80 +1,173 @@
+            if(*count == 0)
+            {
+                delete count;
+                delete m_ptr;
+            }
+            m_ptr = p.m_ptr;
+            count = p.count;
+            p.m_ptr = nullptr;
+            p.count = nullptr;
 
-    SharedPointer(const SharedPointer& sharedPtr){ // Copy constructor
-        this->m_ptr = sharedPtr.m_ptr;
-        this->m_count = sharedPtr.m_count;
-        if(nullptr != sharedPtr.m_ptr)
-        {
-            (*this->m_count)++;
         }
-        
+        return *this;
     }
-    
-    SharedPointer& operator=(const SharedPointer& sharedPtr) //Copy assignment operator
+
+    // Destructor 
+    ~SharedPointer()
     {
-        this->m_ptr = sharedPtr.m_ptr;
-        this->m_count = sharedPtr.m_count;
-        if(nullptr != sharedPtr.m_ptr)
+        (*count)--; 
+        if(*count == 0 )
         {
-            (*this->m_count)++;
+            delete m_ptr;
+            delete count;
         }
     }
     
-    T* operator->() const
-    {
-        return this->m_ptr;
-    }
+    //Dereferencing
+     T& operator *()
+     {
+        return *m_ptr;
+     }
+
+
+     //Pointing
+     T* operator->()
+     {
+        return m_ptr;
+     }
+ 
+     //Return use_count
+     int  use_count() const{
+        return *count;
+     }
     
+    //calculate unique
+     bool unique() const{
+        return *count == 1;
+     }
+
+     //Reset
+     void reset() {
+        *count = *count > 0 ? *count -1: *count;
+        if(*count == 0)
+        {
+            delete m_ptr;
+            delete count;
+        }
+        m_ptr = nullptr;
+        count = new int(1);
+     }
+
+    // Paramatized reset
+     void reset(SharedPointer& new_ptr)
+     {
+        *count = *count > 0 ? *count -1: *count;
+        if(*count == 0)
+        {
+            delete m_ptr;
+            delete count;
+        }
+        m_ptr = new_ptr.m_ptr;
+        count = new_ptr.count;
+        *count++;
+     }
 
-     void cleanup(){
-            cout << "Cleanup is called";
-            (*m_count)--;
-            if (*m_count == 0)
+     //release 
+     T* release()
+     {
+        if(m_ptr)
+        {
+            T* temp_ptr = m_ptr;
+            *count = *count > 0 ? *count -1: *count;
+            if(*count == 0)
             {
-                if (nullptr != m_ptr)
-                    delete m_ptr;
-                delete m_count;
+                delete m_ptr;
+                delete count;
             }
+            m_ptr = nullptr;
+            count = new int(1);
+          return temp_ptr;  
         }
-    
-    int count()
-    {
-        return *m_count;
-    }
+        return nullptr;
 
+     }
 
-    
+    //Swap two function
+     void swap(SharedPointer& other)
+     {
+        T* temp_ptr = m_ptr;
+        T* temp_count = count;
+        m_ptr =  other.m_ptr;
+        count = other.count;
+        other.m_ptr = temp_ptr;
+        other.count = temp_count;
+     }
+    private:
+    T* m_ptr;
```

Comment:
Add a invariant comment here - for both m_ptr and count. And honor that invariant.

Set them nullptr by default directly here , just in case you miss to assign something to them....

Reply:
You didn't fix this yet.

Reply:
fixed